File: crabzilla-vavr-example1/crabzilla-vavr-example1-service/src/main/java/io/github/crabzilla/example1/customer/CustomerFactory.java
package io.github.crabzilla.example1.customer;


import io.github.crabzilla.example1.services.SampleInternalService;
import io.github.crabzilla.model.*;
import io.github.crabzilla.vertx.AggregateRootComponentsFactory;
import io.vertx.core.Vertx;
import io.vertx.ext.jdbc.JDBCClient;

import javax.inject.Inject;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

import static io.github.crabzilla.example1.customer.CustomerFunctions.*;

// tag::factory[]
public class CustomerFactory implements AggregateRootComponentsFactory<Customer> {

  private final Vertx vertx;
  private final JDBCClient jdbcClient;
  private final Customer seedValue;

  @Inject
  public CustomerFactory(SampleInternalService service, Vertx vertx, JDBCClient jdbcClient) {
    this.seedValue = new Customer(service, null, null, false, null);
    this.vertx = vertx;
    this.jdbcClient = jdbcClient;
  }


  @Override
  public Class<Customer> clazz() {
    return Customer.class;
  }

  @Override
  public Supplier<Customer> supplierFn() {
    return () -> seedValue;
  }

  @Override
  public BiFunction<DomainEvent, Customer, Customer> stateTransitionFn() {
    return new StateTransitionFn();
  }

  @Override
  public Function<EntityCommand, List<String>> cmdValidatorFn() {
    return new CommandValidatorFn();
  }

  @Override
  public BiFunction<EntityCommand, Snapshot<Customer>, CommandHandlerResult> cmdHandlerFn() {
    return new CommandHandlerFn(instance ->
            new StateTransitionsTracker<>(instance, stateTransitionFn()));
  }

  @Override
  public JDBCClient jdbcClient() {
    return jdbcClient;
  }

  @Override
  public Vertx vertx() {
    return vertx;
  }

}
// end::factory[]

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-service/src/main/java/io/github/crabzilla/example1/customer/CustomerFactory.kt
package io.github.crabzilla.example1.customer


import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.*
import io.github.crabzilla.example1.SampleInternalService
import io.github.crabzilla.vertx.entity.roles.AggregateComponentsFactory
import io.vertx.core.Vertx
import io.vertx.ext.jdbc.JDBCClient
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Supplier
import javax.inject.Inject
import javax.inject.Singleton

// tag::factory[]
@Singleton
class CustomerFactory @Inject
constructor(service: SampleInternalService, private val vertx: Vertx,
            private val jdbcClient: JDBCClient) : AggregateComponentsFactory<Customer> {

  private val seedValue by lazy { Customer(sampleInternalService = service) }

  override fun clazz(): Class<Customer> {
    return Customer::class.java
  }

  override fun supplierFn(): Supplier<Customer> {
    return Supplier { seedValue }
  }

  override fun stateTransitionFn(): BiFunction<DomainEvent, Customer, Customer> {
    return StateTransitionFn()
  }

  override fun cmdValidatorFn(): Function<EntityCommand, List<String>> {
    return CommandValidatorFn()
  }

  override fun cmdHandlerFn():
          BiFunction<EntityCommand, Snapshot<Customer>, EntityCommandResult> {
    val trackerFactory = StateTransitionsTrackerFactory<Customer> { instance ->
      StateTransitionsTracker(instance, stateTransitionFn())
    }
    return CommandHandlerFn(trackerFactory)
  }

  override fun jdbcClient(): JDBCClient {
    return jdbcClient
  }

  override fun vertx(): Vertx {
    return vertx
  }

}
// end::factory[]
